src/main/java/org/spongepowered/common/item/recipe/crafting/AbstractSpongeCraftingRecipe.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.item.recipe.crafting;
+
+import static org.spongepowered.common.item.inventory.util.InventoryUtil.toSpongeInventory;
+
+import net.minecraft.inventory.InventoryCrafting;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipe;
+import net.minecraft.util.NonNullList;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.item.inventory.type.GridInventory;
+import org.spongepowered.api.item.recipe.crafting.CraftingRecipe;
+import org.spongepowered.api.world.World;
+import org.spongepowered.common.item.inventory.util.ItemStackUtil;
+
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public abstract class AbstractSpongeCraftingRecipe implements CraftingRecipe, IRecipe {
+    @Override
+    public boolean matches(InventoryCrafting inv, net.minecraft.world.World worldIn) {
+        return matches(this::isValid, inv, worldIn);
+    }
+
+    public static boolean matches(BiFunction<GridInventory, World, Boolean> isValid, InventoryCrafting inv, net.minecraft.world.World worldIn) {
+        return isValid.apply(toSpongeInventory(inv), (World) worldIn);
+    }
+
+    @Override
+    public ItemStack getCraftingResult(InventoryCrafting inv) {
+        return getCraftingResult(this::getResult, inv);
+    }
+
+    public static ItemStack getCraftingResult(Function<GridInventory, ItemStackSnapshot> getResult, InventoryCrafting inv) {
+        return ItemStackUtil.fromSnapshotToNative(getResult.apply(toSpongeInventory(inv)));
+    }
+
+    @Override
+    public int getRecipeSize() {
+        return getRecipeSize(this::getSize);
```

Comment:
The pattern is just IoC but that simple fact is being occluded by the actual method of implementation. I had the same problem with `Adapter` but solved it in a slightly clearer way of abstracting the logic away to an inner class `Adapter.Logic` and then any *Trait* mixins call back against the static methods therein.

Honestly for consistency a similar approach should be used here. Instead of static methods in a base class, wrap the static methods in a `Logic` or `Operations` class and call back against it. The mixins should also probably be *Trait* in this case to denote their usage.